openssl/hmac.go

```diff
@@ -146,3 +164,90 @@ func hmacCtxFree(ctx C.GO_HMAC_CTX_PTR) {
 	}
 	C.go_openssl_HMAC_CTX_free(ctx)
 }
+
+// hmac3 implements hash.Hash
+// using functions available in OpenSSL 3.
+type hmac3 struct {
+	md        C.GO_EVP_MAC_PTR
+	ctx       C.GO_EVP_MAC_CTX_PTR
+	params    [2]C.OSSL_PARAM
+	size      int
+	blockSize int
+	key       []byte
+	sum       []byte
+}
+
+func newHMAC3(key []byte, h hash.Hash, md C.GO_EVP_MD_PTR) *hmac3 {
+	mac := C.go_openssl_EVP_MAC_fetch(nil, paramAlgHMAC, nil)
+	ctx := C.go_openssl_EVP_MAC_CTX_new(mac)
+	if ctx == nil {
+		panic("openssl: EVP_MAC_CTX_new failed")
+	}
+	digest := C.go_openssl_EVP_MD_get0_name(md)
+	params := [2]C.OSSL_PARAM{
+		C.go_openssl_OSSL_PARAM_construct_utf8_string(paramDigest, digest, 0),
+		C.go_openssl_OSSL_PARAM_construct_end(),
+	}
+	if C.go_openssl_EVP_MAC_init(ctx, base(key), C.size_t(len(key)), &params[0]) == 0 {
+		panic(newOpenSSLError("EVP_MAC_init failed"))
```

Comment:
I forget what we decided about freeing memory vs. `panic`s... would it make sense to shuffle stuff around so if this panic happens and the caller recovers, we finalize? Something like this:

```go
	ctx := C.go_openssl_EVP_MAC_CTX_new(mac)
	if ctx == nil {
		panic("openssl: EVP_MAC_CTX_new failed")
	}
	hmac := &hmac3{
		md:        mac,
		ctx:       ctx,
		size:      h.Size(),
		blockSize: h.BlockSize(),
		key:       key,
	}
	runtime.SetFinalizer(hmac, (*hmac3).finalize)
	digest := C.go_openssl_EVP_MD_get0_name(md)
	params := [2]C.OSSL_PARAM{
		C.go_openssl_OSSL_PARAM_construct_utf8_string(paramDigest, digest, 0),
		C.go_openssl_OSSL_PARAM_construct_end(),
	}
	hmac.params = params
	if C.go_openssl_EVP_MAC_init(ctx, base(key), C.size_t(len(key)), &params[0]) == 0 {
		panic(newOpenSSLError("EVP_MAC_init failed"))
	}
	return hmac
```

Reply:
Good catch. I've updated the code to call `hmac.Reset()` at the bottom of the function, which will call `go_openssl_EVP_MAC_init`. This way we also remove some duplication.